### PR TITLE
Formatting combinators

### PR DESCRIPTION
**Pull request by AJChapman**
This adds formatting combinators -- functions which take a Format and produce a new, modified Format. Also includes tests and haddocks for these.

I also have some other random commits improving other things. Happy to break them out into separate commits if you like.

### CODE REVIEW

File: src/Formatting/Combinators.hs [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+
+-- | Format each value in a list and concatenate them all:
+--
+-- >>> format (concatenated text) ["one", "two", "three"]
+-- "onetwothree"
+--
+-- >>> format (took 15 (concatenated bin)) [1..]
+-- "1101110010111011110001001101010111100110111101111"
+concatenated :: Foldable t => Format Builder (a -> Builder) -> Format r (t a -> r)
+concatenated f = later $ toList
```
Comment (endgame) [inline src/Formatting/Combinators.hs:97]: Is there a simpler expression using `foldMap`?

Comment (AJChapman) [reply src/Formatting/Combinators.hs]: Yes! I'll update that.

File: src/Formatting/Combinators.hs [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+-- makeLenses ''Person
+--
+-- me :: Person
+-- me = Person "Alex" 38
+--
+-- format ("The person's name is " % quoted (viewed personName text) % ", and their age is " <> viewed personAge int) me
+-- "The person's name is 'Alex', and their age is 38"
+-- @
+viewed :: ((a -> Const a b) -> s -> Const a t) -> Format r (a -> r) -> Format r (s -> r)
+viewed l = fmap (. (getConst . l Const))
```
Comment (axman6-da) [inline src/Formatting/Combinators.hs:513]: maybe also add a non-lens `view` or `viewing` function here with an accessor function argument? (I'm sure this exists, but it seems like a consistent thing to have)

File: test/Spec.hs [hunk 3]
```diff
@@ -123,3 +124,59 @@ spec = do
     it "hour" $ flip shouldBe "00:01:00:00" $ format diffComponents 3600
     it "day" $ flip shouldBe "01:00:00:00" $ format diffComponents 86400
 
+  describe "list formatters" $ do
+    it "concatenated" $ flip shouldBe "onetwothree" $ format (concatenated text) ["one", "two", "three"]
```
Comment (axman6-da) [inline test/Spec.hs:128]: Does the precedence of the operators allow this to be 
```suggestion
    it "concatenated" $ format (concatenated text) ["one", "two", "three"] `shouldBe` "onetwothree"
```
or would it need 
```suggestion
    it "concatenated" (format (concatenated text) ["one", "two", "three"] `shouldBe` "onetwothree")
```
(which isn't too gross IMO)

**Review by axman6-da — COMMENTED** [review]
Love these combinators, and excellent documentation. One thing that I have implemented many times myself are fixed width hex and binary formatters; `format (binPrefix 16) 5 -> "0b0000000000000101`. I'd love to see those added too.

File: src/Formatting/Combinators.hs [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+worded
+  :: (Format r' (Builder -> r') -> Format Builder ([Builder] -> Builder)) -- ^ A list-formatting combinator, e.g. 'unworded', 'list', 'concatenated', etc.
+  -> Format r a -- ^ The base formatter, whose rendered text will be split
+  -> Format r a
+worded = splatWith TL.words
+
+-- | Split the formatted item into lines and use the given list combinator to render the resultant list of strings.
+--
+-- >>> format (lined list text) "one two three\n\nfour five six\nseven eight nine\n\n"
+-- "[one two three, , four five six, seven eight nine, ]"
```
[the remark this answers is not in the document]
Comment (AJChapman) [reply src/Formatting/Combinators.hs:233]: There's a single blank line at the end of the input, hence the extra empty string at the end of the list. I'll make this more obvious...

File: src/Data/Text/Format/Types.hs [hunk 5]
```diff
@@ -19,11 +19,6 @@ module Data.Text.Format.Types
     , Hex(..)
     ) where
 
-import Data.Monoid (Monoid(..))
-import Data.String (IsString(..))
-import Data.Text (Text)
-import Data.Typeable (Typeable)
-
```
[the remark this answers is not in the document]
Comment (AJChapman) [reply src/Data/Text/Format/Types.hs:26]: Oh... I think not, but I can't be sure without testing...

**Pull request merged**